### PR TITLE
#912 allow leading period

### LINKED ISSUE #912: Support file paths with a dotted basename
Currently, I think the logic tries to set the "derivation" name to the basename when copying a path into the store. If that basename starts with a dot, Nix moans about not being able to copy it in.

Luckily, the derivation name doesn't matter all that much in Nix, so it can really be anything we want. Perhaps we could just have the file copying code strip out any leading dots from the name, or replace them with the string "dot"? That way we could do something like this:

``` nix
{
  foo = mkDerivation {
    name = "foo";
    src = ./foo/bar/.baz;
  };
}
```

### PR DESCRIPTION
**Pull request by roberth**
# Motivation

#9095 seems to have caused more harm than good. See https://github.com/NixOS/nix/pull/9095#issuecomment-1888097127 for argumentation.

- Closes #912

# Context

- #9091
- #9095
- #9224
- #9219
- #9130
- #9120
- #9121
- #9122

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

**Comment by edolstra** [comment]
I think this sets a bad precedent of having to support bugs that accidentally snuck in forever. Also, flip-flopping on this behaviour *again* just makes the situation even more confusing.

**Comment by roberth** [comment]
> bug

It was requested as a feature, #912

> a bad precedent of having to support bugs that accidentally snuck in forever.

What if denying `.` is the bug?

> a bad precedent

Since when do we make decisions based on precedent? 
You're basically suggesting that we're just imitating the past, and we're not allowed to correct our mistakes?

The author of #9095 has shown no interest in helping users with the consequences of their PR. I don't blame them for that; they can have all sorts of reasons, and it's a risk I took when accepting their PR.
However, unlike them, I _am_ a maintainer, and that makes me responsible for this issue by default.

If I were to hand responsibility to you, how would you deal with the problem? We have users with broken stores now, who can't upgrade their Nix. Do you plan to fix the problem in a different way? What about #912?

**Comment by roberth** [comment]
@Ericson2314 this fixes two Arbitrary instances. Some more still remain.

**Comment by Ericson2314** [comment]
Glad to hear it!

File: doc/manual/rl-next/leading-period.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+synopsis: Store paths are allowed to start with `.`
+issues: 912
+prs: 9867 9091 9095 9120 9121 9122 9130 9219 9224
+---
+
+Leading periods were allowed by accident in Nix 2.4. The Nix team has considered this to be a bug, but this behavior has since been relied on by users, leading to unnecessary difficulties.
+From now on, leading periods are officially, definitively supported. The names `.` and `..` are disallowed, as well as those starting with `.-` or `..-`.
+
+Nix versions that denied leading periods are
+- 2.19
```
Comment (Ericson2314) [inline doc/manual/rl-next/leading-period.md:11]: N.B. Not 2.20 because backport.

File: src/libstore/path-regex.hh [hunk 2]
```diff
@@ -3,6 +3,6 @@
 
 namespace nix {
 
-static constexpr std::string_view nameRegexStr = R"([0-9a-zA-Z\+\-_\?=][0-9a-zA-Z\+\-\._\?=]*)";
+static constexpr std::string_view nameRegexStr = R"((?!\.\.?(-|$))[0-9a-zA-Z\+\-\._\?=]+)";
```
Comment (Ericson2314) [inline src/libstore/path-regex.hh:6]: I don't quite understand this one. Maybe it needs some comments?

File: src/libstore/path.cc [hunk 3]
```diff
@@ -9,9 +9,20 @@ static void checkName(std::string_view path, std::string_view name)
...
             path, StorePath::MaxPathLen);
-    if (name[0] == '.')
-        throw BadStorePath("store path '%s' starts with illegal character '.'", path);
     // See nameRegexStr for the definition
+    if (name[0] == '.') {
+        // check against "." and "..", followed by end or dash
+        if (name.size() == 1)
+            throw BadStorePath("store path '%s' has invalid name '%s'", path, name);
+        if (name[1] == '-')
+            throw BadStorePath("store path '%s' has invalid name '%s': first dash-separated component must not be '%s'", path, name, ".");
```
Comment (Ericson2314) [inline src/libstore/path.cc:18]: Could pull this out into a lambda (with param for `.` or `..`) and use twice

**Review by Ericson2314 — APPROVED** [review]
Looks good! None of my nits be low I care very much about :)

File: doc/manual/rl-next/leading-period.md (hunk 1 above)
Comment (roberth) [reply doc/manual/rl-next/leading-period.md:11]: Turn into link to issue comment.

File: src/libstore/path-regex.hh (hunk 2 above)
Comment (roberth) [reply src/libstore/path-regex.hh:6]: Done.

**Comment by thufschmitt** [comment]
(bit late, but): Discussed during the Nix maintainers meeting on 2024-01-26.
Agreement to merge with a few extra restrictions (forbid things like `.` or `..`).

<details>
- Path with a leading `.` used to be forbidden, then got allowed by accident, and forbidden again
    - Problematic because users might now have invalid paths in their store, that Nix can't do anything with
    - Suggestion by @roberth: Allow them for good
- @edolstra: Is it just a matter of a few people having them in their store? Because that can be worked around
- @thufschmitt: What are the cost of allowing that for good?
    - @roberth: 2.3 doesn't support them
    - Might be security-relevant?
        - Dotfiles being hidden, this might be relied on for security reasons
        - Example: Some nixpkgs code somewhat relying on things not starting with a dot
- @edolstra: These names are meant to be package names, not arbitrary ones (and we actually forbid a lot of characters already). What would it mean to have a package starting with a `.`?
- @roberth: Nix is also used as a config manager. Shouldn't we use it for that?
- @edolstra: Is there any package manager that allows package names starting with a `.`
    - @tom: At least RPM does
- @thufschmitt: If we don't allow these, how do make sure that people who already have them don't get bitten?
    - No real solution except allowing them internally (just forbid creating new ones)
    
- Need to forbid things like `.`, `..`, `..-`* and the like
    - hopefully reasonable restriction: Require at least one alphanumeric character before the `a`
    - also update the #5380 copy of the regex
    
Agreement

**Comment by ryantm** [comment]
Did this get rolled back? I'm seeing
```
error: store path '1gzd3gdn5ff8ndinqdxsbpbw08yzcl87-.gitignore' starts with illegal character '.'
```
when running `sudo nix-collect-garbage -d`
in nix 2.18.2 and nix 2.21.0

**Comment by roberth** [comment]
@ryantm It didn't get reverted. Maybe you're getting this from a daemon or remote builder?

**Comment by ryantm** [comment]
@roberth Thanks for the quick reply, that was a good hint. I had:

```
$ nix store ping
Store URL: daemon
Version: 2.18.2
Trusted: 1
```
then I updated my NixOS nix to be
```
$ nix store info
Store URL: daemon
Version: 2.21.0
Trusted: 1
```
and garbage collection worked fine.

**Comment by thufschmitt** [comment]
This doesn't seem to have been backported to 2.18? We need to

**Comment by edef1c** [comment]
> The author of #9095 has shown no interest in helping users with the consequences of their PR. I don't blame them for that; they can have all sorts of reasons, and it's a risk I took when accepting their PR.
> However, unlike them, I _am_ a maintainer, and that makes me responsible for this issue by default.

To be clear, my initial proposal (#9091) was to document and codify the change in behaviour, rather than revert it, given that it was likely to have snuck into real-world use already, as experience has now indeed borne out.

I can't spend all my time glued to GitHub currently, so this is the first time I'm seeing the fallout. It seemed like a perhaps risky but worthwhile experiment, that you [endorsed](https://github.com/NixOS/nix/pull/9091#issuecomment-1746607430) as a maintainer (before code for the stricter validation was even written), and I'm fine with how this has concluded.

We don't have an "ecosystem test run" environment like Rust's [Crater](https://rustc-dev-guide.rust-lang.org/tests/crater.html), so I would not (and could not) have undertaken a live-fire test like this without a maintainer taking responsibility for it. Building a Crater equivalent seems abstractly worthwhile, but right now I don't have the bandwidth for such an undertaking, and I don't particularly expect anyone to volunteer.

As far as I can tell, the only alternative unexplored was making it a warning and treating it as an explicitly deprecated behaviour, but I don't think that would have done much good.

My sole goal here was to ensure that there is a clear agreement on what constitutes in-spec Nix behaviour, so that other implementers (such as the TVL nix-compat Rust library, which I co-maintain) can match it in behaviour exactly, since we lack a formal specification.

**Comment by roberth** [comment]
Hi @edef1c and thank you for your response.
We all did the best we could do with our limited resources.

You're making good points about the ecosystem test run and a formal specification. We're trying to improve both, but for the formal specification, that takes a different form. Making a separate spec document is an effort that's hard to justify (for the Nix team), but we are making the documentation more precise. A spec is also reference documentation, so I think there's a lot that can be done about this issue, just by improving the Nix manual. So, contributions are welcomed; especially small and incremental ones.

**Comment by vcunat** [comment]
I believe this hasn't been applied to 2.19 (yet).  Though 9ddd0f2af8fd95e1380027a70d0aa650ea2fd5e4 would at least apply cleanly.

**Comment by roberth** [comment]
We don't actively support 2.19, but feel free to open a PR

**Comment by vcunat** [comment]
Sad that the current Hydra is based on 2.19.

**Pull request merged**